core/backend/src/test/sheetindex/SheetIndex.test.ts

```diff
@@ -0,0 +1,149 @@
+ *--------------------------------------------------------------------------------------------*/
+
+import { Id64, Id64String } from "@itwin/core-bentley";
+import { BisCodeSpec, CodeScopeSpec, CodeSpec, GeometricModel2dProps, RelatedElement, SheetProps } from "@itwin/core-common";
+
+import { IModelDb, SnapshotDb } from "../../IModelDb";
+import { ExtensiveTestScenario, IModelTestUtils } from "../IModelTestUtils";
+import { Sheet, SheetIndex, SheetIndexFolder, SheetReference } from "../../Element";
+import { expect } from "chai";
+import { DocumentListModel, SheetIndexModel, SheetModel } from "../../Model";
+
+let documentListModelId: string | undefined;
+const getOrCreateDocumentList = async (iModel: IModelDb): Promise<Id64String> => {
+  // If they do not exist, create the document partition and document list model
+  if (documentListModelId === undefined) {
+    const channelSubjectId = iModel.elements.getRootSubject().id;
+    await iModel.locks.acquireLocks({
+      shared: channelSubjectId,
+    });
+    documentListModelId = DocumentListModel.insert(iModel, channelSubjectId, "SheetsList");
+  }
+
+  return documentListModelId;
+};
+
+const insertSheet = async (iModel: IModelDb, sheetName: string): Promise<Id64String> => {
+  const createSheetProps = {
+    height: 42,
+    width: 42,
+    scale: 42,
+  };
+  // Get or make documentListModelId
+  const modelId = await getOrCreateDocumentList(iModel);
+
+  // Acquire locks and create sheet
+  await iModel.locks.acquireLocks({ shared: documentListModelId });
+  const sheetElementProps: SheetProps = {
+    ...createSheetProps,
+    classFullName: Sheet.classFullName,
+    code: Sheet.createCode(iModel, modelId, sheetName),
+    model: modelId,
+  };
+  const sheetElementId = iModel.elements.insertElement(sheetElementProps);
+
+  const sheetModelProps: GeometricModel2dProps = {
+    classFullName: SheetModel.classFullName,
+    modeledElement: { id: sheetElementId, relClassName: "BisCore:ModelModelsElement" } as RelatedElement,
+  };
+  const sheetModelId = iModel.models.insertModel(sheetModelProps);
+
+  return sheetModelId;
+};
+
+const insertCodeSpec = async (iModel: IModelDb) => {
+  const indexSpec = CodeSpec.create(iModel, BisCodeSpec.sheetIndex, CodeScopeSpec.Type.Model);
+  iModel.codeSpecs.insert(indexSpec);
+
+  const entrySpec = CodeSpec.create(iModel, BisCodeSpec.sheetIndexEntry, CodeScopeSpec.Type.ParentElement);
+  iModel.codeSpecs.insert(entrySpec);
+};
+
+describe("SheetIndex", () => {
+  let iModel: SnapshotDb;
+
+  beforeEach(async () => {
+    const iModelFile: string = IModelTestUtils.prepareOutputFile("IModel", "TestSheetIndex.bim");
+
+    const iModelDb = SnapshotDb.createEmpty(iModelFile, { rootSubject: { name: "SheetIndex" } });
+    await ExtensiveTestScenario.prepareDb(iModelDb);
+    ExtensiveTestScenario.populateDb(iModelDb);
+    iModel = iModelDb;
+
+    await insertCodeSpec(iModel);
```

Comment:
Is this necessary?  This steps feel like it should already have happened.  Is this what was meant by, "Ensure the new codes are included in pre-existing iModels"?

Reply:
I'll defer to @aruniverse to investigate and confirm this question.

Reply:
Built-in code specs are registered in native code when creating a new DgnDb by [DgnDb::CreateCodeSpecs](https://github.com/iTwin/imodel-native/blob/f6244d40dd38146680bbdd0f7fc711b6a3f7f02c/iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp#L287).
That function notably does not register sheet index-related code specs.

Reply:
CodeSpecs use server-issued Ids. There is [some attempt](https://github.com/iTwin/itwinjs-core/blob/10d8e52c046250c8650688a84bf9a196c978b722/core/backend/src/CodeSpecs.ts#L102) to prevent conflicts between briefcases. I think you will need to make your `createCode` functions ensure that they register the relevant CodeSpec if not already present in the iModel. You may wish to add documentation to your new BisCodeSpec members indicating that they are only created in an iModel on first use.

@khanaffan should we consider a DgnDb schema upgrade that adds these code specs if not present?

Reply:
Updating the iModel indirectly seems like a side effect we should avoid, but I'm not sure if there's an instance where a user would want to create a code an not immediately update the iModel with the next element.  Plus, I can't think of a more elegant way.

Implemented as suggested

Reply:
Yes we can.

Reply:
Though feel like we should have different approach where its like define in custom attribute in BisCore. So we do not have to add them via profile version change instead biscore can be modified to add builtin code that will exist in seed file or upgraded file. So we can tie it to biscore rather then runtime.

Reply:
@khanaffan can you ping me sometime today to clarify your suggestion please?

Reply:
After discussing with @khanaffan, it was decided that the suggest implementation should be reverted.

Currently, these added classes are all in beta and will main be used by my team.  Once they have matured a bit and are ready to be released, the new Code Specs will be added to the profile and method of updating existing iModels will the added then.  In the mean time, the exact way to add them will be discussed more.